fabric-api-base/src/main/java/net/fabricmc/fabric/api/event/EventFactory.java

```diff
@@ -47,18 +47,70 @@ public static void invalidate() {
 
 	/**
 	 * Create an "array-backed" Event instance.
+	 * The factory will be used for any number of listeners.
+	 * Consider using {@link #createArrayBackedCustomEmptyInvoker(Class, Object, Function)}  the more optimized overload} if you need a slight performance increase
+	 * when there are 0 or 1 listeners.
 	 *
 	 * @param type           The listener class type.
 	 * @param invokerFactory The invoker factory, combining multiple listeners into one instance.
 	 * @param <T>            The listener type.
 	 * @return The Event instance.
 	 */
+	public static <T> Event<T> createArrayBackedUsingTheFactoryEveryTime(Class<? super T> type, Function<T[], T> invokerFactory) {
+		return EventFactoryImpl.createArrayBackedUsingTheFactoryEveryTime(type, invokerFactory);
+	}
+
+	/**
+	 * Create an "array-backed" Event instance, with a custom empty invoker.
+	 * The empty invoker will be used when there are no listeners,
+	 * and when there is only one listener, it will be used directly.
+	 * The factory will only be used when there are at least two listeners.
+	 * Consider using {@link #createArrayBackedUsingTheFactoryEveryTime} if this optimization is unsuitable or unneeded.
+	 *
+	 * @param type           The listener class type.
+	 * @param emptyInvoker   The custom empty invoker.
+	 * @param invokerFactory The invoker factory, combining multiple listeners into one instance.
+	 * @param <T>            The listener type.
+	 * @return The Event instance.
+	 */
+	public static <T> Event<T> createArrayBackedCustomEmptyInvoker(Class<? super T> type, T emptyInvoker, Function<T[], T> invokerFactory) {
+		return createArrayBackedUsingTheFactoryEveryTime(type, invokers -> {
+			if (invokers.length == 0) {
+				return emptyInvoker;
+			} else if (invokers.length == 1) {
+				return invokers[0];
+			} else {
+				return invokerFactory.apply(invokers);
+			}
+		});
+	}
+
+	/**
+	 * Create an "array-backed" Event instance.
+	 * The factory will be used when there are zero listeners, or at least two.
+	 * If there is only one listener, that one will be used as the invoker, and the factory will not be used!
+	 *
+	 * @param type           The listener class type.
+	 * @param invokerFactory The invoker factory, combining multiple listeners into one instance.
+	 * @param <T>            The listener type.
+	 * @return The Event instance.
+	 * @deprecated Use {@link #createArrayBackedUsingTheFactoryEveryTime(Class, Function) createUnbad}.
+	 */
+	@Deprecated
 	public static <T> Event<T> createArrayBacked(Class<? super T> type, Function<T[], T> invokerFactory) {
-		return EventFactoryImpl.createArrayBacked(type, invokerFactory);
+		return createArrayBackedUsingTheFactoryEveryTime(type, invokers -> {
```

Comment:
No, you can just pass in the old invoker factory. The old invoker factory doesn't break when the input array is empty or has length 1. So you can just replace the old `createArrayBacked` with `createArrayBackedUsingTheFactoryEveryTime` without concerns

Reply:
I am hesitant to introduce such a subtle behavior change. What if people relied on the old behavior? (I don't know why, but still)